File: backend/src/api/add.js
// file: src/api/add.js

const fs = require('fs');
const crypto = require('crypto');
const util = require('util');
const exec = util.promisify(require('child_process').exec);

/**
 * Executes user's add function in a supported language (javascript, python).
 * 
 * @param {*} req - expects body in form { a: 1, b: 2, language: "javascript", code: "function add(a, b) { ... }" }
 * @param {*} res - returns body in form { c: 3, output: " ... " }
 */
async function add(req, res) {
  // print status message
  console.log('[ STATUS ] POST /add -', Date());
  //console.log('request body:', req.body);

  // check for missing body fields
  if (!req.body.a || !req.body.b || !req.body.language || !req.body.code) {
    res.setHeader('Content-Type', 'application/json');
    res.status(400);
    res.send(JSON.stringify({ "error": "missing fields in body" }));
    return;
  }

  // generate unique string
  const unique = crypto.randomBytes(10).toString('hex');



  // run user code for specified language
  let stdout, stderr, fileName, contents;
  switch (req.body.language) {
    case 'javascript':
      // write user code to unique filename
      fileName = `${unique}.js`;
      contents = [
        req.body.code,
        `const c = add(${req.body.a}, ${req.body.b});`,
        `console.log('--- ${unique} ---');`,
        'console.log(c);',
      ].join('\n');
      fs.writeFileSync(fileName, contents);

      // attempt to run user code and record output
      try {
        ({ stdout, stderr } = await exec(`node ${fileName}`));
      } catch (error) {
        console.log('[ ERROR ] an error occured while executing user\'s code:\n', error);
        stderr += error;
      }

      // remove user code file
      fs.unlinkSync(fileName);
      break;

    case 'python':
      // write user code to unique filename
      fileName = `${unique}.py`;
      contents = [
        req.body.code,
        `c = add(${req.body.a}, ${req.body.b})`,
        `print('--- ${unique} ---');`,
        'print(c);',
      ].join('\n');
      fs.writeFileSync(fileName, contents);

      // attempt to run user code and record output
      try {
        ({ stdout, stderr } = await exec(`python3 ${fileName}`));
      } catch (error) {
        console.log('[ ERROR ] an error occured while executing user\'s code:\n', error);
        stderr += error;
      }

      // remove user code file
      fs.unlinkSync(fileName);
      break;

    // if langauge did not match an option, send error
    default:
      res.setHeader('Content-Type', 'application/json');
      res.status(400);
      res.send(JSON.stringify({ "error": `unsupported language (${req.body.language})` }));
      return;
  }

  // separate program output from result
  if (!stdout) stdout = '';
  const splitStdout = stdout.split(`--- ${unique} ---\n`)
  const output = ['--- stdout ---', splitStdout[0], '--- stderr ---', stderr].join('\n');

  // extract result
  const c = parseInt(splitStdout[1]);

  // send response
  res.setHeader('Content-Type', 'application/json');
  res.send(JSON.stringify({ 'c': c, 'output': output }));
}

module.exports = add;

//
// end of file: src/api/add.js

File: backend/src/api/add.go
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"

	f "subparoprogramming.org/algosim-backend/funcs"
)

type AddRequest struct {
	A        string
	B        string
	Language string
	Code     string
}

type AddResponse struct {
	C      int    `json:"c"`
	Output string `json:"output"`
}

// handler function for the /add endpoint
func Add(w http.ResponseWriter, r *http.Request) {
	fmt.Printf("[ STATUS ] request recieved (%s) %v\n", r.URL.Path, time.Now())

	f.EnableCors(&w)

	// only process POST requests
	if r.Method != "POST" {
		fmt.Fprintf(w, "{ \"error\": \"only POST method is supported\" }")
		return
	}

	var reqBody AddRequest

	// Try to decode the request body into the struct. If there is an error,
	// respond to the client with the error message and a 400 status code.
	err := json.NewDecoder(r.Body).Decode(&reqBody)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	unique := f.RandomHexString(20)

	var filename, contents string
	var cmd *exec.Cmd

	switch reqBody.Language {
	case "javascript":
		// create filename and file contents
		filename = unique + ".js"
		contents = reqBody.Code + `
const c = add(` + reqBody.A + `, ` + reqBody.B + `);
console.log('--- ` + unique + ` ---');
console.log(c);`

		// create command
		cmd = exec.Command("node", filename)

	case "python":
		// create filename and file contents
		filename = unique + ".py"
		contents = reqBody.Code + `
c = add(` + reqBody.A + `, ` + reqBody.B + `)
print('--- ` + unique + ` ---')
print(c)`

		// create command
		cmd = exec.Command("python3", filename)

	default:
		http.Error(w, "{ \"error\": \"unrecognized language\" }", http.StatusBadRequest)
		return
	}

	// write user code to executable file
	err = os.WriteFile(filename, []byte(contents), 0777)
	f.CheckError(err)

	// execute file and capture stdout and stdin
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stdout
	err = cmd.Run()

	// delete file
	os.Remove(filename)

	// setparate program output from result
	splitStdout := strings.Split(stdout.String(), "--- "+unique+" ---\n")
	output := `--- stdout ---
` + splitStdout[0] + `
--- stderr ---
` + stderr.String()

	// extract result
	var c int
	if len(splitStdout) > 1 {
		c, _ = strconv.Atoi(strings.Split(splitStdout[1], "\n")[0])
		f.CheckError(err)
	} else {
		c = -1
	}

	// format and send response
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(AddResponse{C: c, Output: output})
}
